### PR TITLE
#7 product manager class

### LINKED ISSUE #7: Create productManager
Create a class which will manage product operations like:
- deleteProduct
- createProduct
- getProductsBySku
- getProductById
- archiveProduct
- deleteVariantFromProduct
- isArchivedProduct

### PR DESCRIPTION
**Pull request by junajan**
This PR contains productManager class which has methods for modifying product on ctp API
Resolves #7

### CODE REVIEW

File: lib/services/product-manager.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    return this.client.products
+      .byId(product.id)
+      .delete(product.version)
+  }
+
+  /**
+   * Function accepts product with staged and current version and a list of SKUs
+   * which will be deleted from a product.
+   * @param product Product object
+   * @param skus List of SKUs
```
Comment (lojzatran) [inline lib/services/product-manager.js]: Can you document how the final actions should be ordered? E.g. `['delete product'] or ['unpublish', 'addVariant', 'removeVariant']...`

File: lib/services/transaction-manager.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import Promise from 'bluebird'
+import * as constants from '../constants'
+
+export default class TransactionManager {
```
Comment (lojzatran) [inline lib/services/transaction-manager.js]: Maybe a comment about what a transaction means in this case would be good.

File: lib/services/transaction-manager.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  constructor (logger, client) {
+    this.client = client
+    this.logger = logger.child({ service: 'transactionManager' })
+
+    this.transactionContainer = constants.TRANSACTION_CONTAINER
```
Comment (lojzatran) [inline lib/services/transaction-manager.js]: Why do you need to assign to field var and not using the constant directly?

File: spec/integration/product-manager.spec.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
+import _ from 'lodash'
```
Comment (lojzatran) [inline spec/integration/product-manager.spec.js]: Let's not create integration tests for every class. They should really be blackbox integration testing where you call the main method of Variant reassignment and then you check the result of the whole reassignment process. So it means we will have only integration tests created from the wiki, which should cover all the methods we have. For the services, you can create unit classes, it should be enough.

File: spec/integration/transaction-manager.spec.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import _ from 'lodash'
```
Comment (lojzatran) [inline spec/integration/transaction-manager.spec.js]: Same here, we don't need integration test. They takes time and resources to run as they need to call CT platform.

File: .eslintignore [hunk 6]
```diff
@@ -4,4 +4,5 @@ dist
 node_modules
 logs
 coverage
-examples
\ No newline at end of file
+examples
+.nyc_output
```
[the remark this answers is not in the document]
Comment (junajan) [reply .eslintignore]: This is the output of [nyc](https://github.com/istanbuljs/nyc) command which is used for generating coverage reports.

File: .npmignore [hunk 7]
```diff
@@ -0,0 +1,19 @@
+.DS_Store
```
[the remark this answers is not in the document]
Comment (junajan) [reply .npmignore]: Ahh... to be honest, I did not know it uses gitignore in case of missing npmignore .. thank you :-) 3a397b5

File: .travis.yml [hunk 8]
```diff
@@ -9,6 +9,7 @@ cache:
 install:
   - npm install -g codecov
   - npm install
+  - npm install -g codecov
```
[the remark this answers is not in the document]
Comment (junajan) [reply .travis.yml]: Result of merging different branches .. removed 7b0f70b

File: lib/services/product-manager.js [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+  createProduct (product) {
+    this.logger.debug('Creating product', product)
+
+    return this.client.products
+      .create(product)
+      .then(res => res.body)
+  }
+
+  publishProduct (product) {
+    const actions = [{
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: Added here 7b0f70b

File: lib/services/product-manager.js [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  }
+
+  filterOutDuplicateProducts (products) {
+    console.log(products)
+    throw new Error('NOT IMPLEMENTED')
+  }
+
+  async getProductsBySkus (skus) {
+    // filter out duplicates or cast to array if needed
+    skus = _.isArray(skus) ? _.uniq(skus) : [skus]
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: You are right .. my intention was to allow also single sku to be given as a parameter but lets make it more strict 7b0f70b

File: lib/services/product-manager.js [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+
+    // if skus have more than a certain amount run multiple parallel requests
+    if (skus.length > this.loadBatchCount) {
+      const productBatches = Promise.map(
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: 👍 thx for catching this a98fcd1

File: lib/services/product-manager.js [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+        _.chunk(skus, this.loadBatchCount), // load max N skus per one request
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: Updated here a98fcd1

File: lib/services/product-manager.js [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+        skuBatch => this.getProductsBySku(skuBatch),
+        { concurrency: this.loadConcurrency } // load products with concurrency
+      )
+
+      return this.filterOutDuplicateProducts(_.flatten(productBatches))
+    }
+
+    const skuPredicate = skus.join('","')
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: Updated here 449906a

File: lib/services/product-manager.js [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+
+  getProductById (id) {
+    return this.client.productProjections
+      .staged(true)
+      .byId(id)
+      .fetch()
+      .then(res => res.body)
+      .catch(err => (
+        err && err.body && err.body.statusCode
+          ? Promise.resolve(undefined)
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: @lojzatran what is the better approach? When the product was not found we should return `null/undefined` or we should throw an exception?

[the remark this answers is not in the document]
Comment (lojzatran) [reply lib/services/product-manager.js]: Probably you missed `err.body.statusCode === 404` :)

File: lib/services/product-manager.js [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+    // first remove variants from staged
+    ['staged', 'current'].forEach((version) => {
+      const variantMap = this._getVariantsBySkuMap(masterData[version])
+
+      // remove variants specified in skus param
+      // but only if product has this variant
+      skus.forEach((sku) => {
+        if (variantMap[sku])
+          actions.push(
+            this._getRemoveVariant(variantMap[sku], version === 'staged')
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: Updated 3693105

File: lib/services/product-manager.js (hunk 1 above)
Comment (junajan) [reply lib/services/product-manager.js]: Documentation added here 3693105

File: lib/services/transaction-manager.js (hunk 3 above)
Comment (junajan) [reply lib/services/transaction-manager.js]: I think that the constants can be sometimes passed as object properties so they can be modified in integration tests. But not this one so I removed it dc44255

File: spec/integration/product-manager.spec.js (hunk 4 above)
Comment (junajan) [reply spec/integration/product-manager.spec.js]: Hello @lojzatran please read my opinion on service integration tests here https://github.com/commercetools/commercetools-node-variant-reassignment/pull/9#discussion_r154577140

With integration tests only for variant reassignment process, you won't cover a lot of different edge cases in service methods. Also, you will test a lot of parts repeatedly (eg. fetching products...) just to test different behavior in one method so the tests will do a lot of useless steps which takes time. I created this issue in which we should remove redundant tests https://github.com/commercetools/commercetools-node-variant-reassignment/issues/16

File: spec/integration/transaction-manager.spec.js (hunk 5 above)
Comment (junajan) [reply spec/integration/transaction-manager.spec.js]: See my previous comment https://github.com/commercetools/commercetools-node-variant-reassignment/pull/15#discussion_r154595587

File: lib/services/product-manager.js (hunk 1 above)
Comment (lojzatran) [reply lib/services/product-manager.js]: So complex just to remove a variant :D

File: lib/services/transaction-manager.js (hunk 2 above)
Comment (junajan) [reply lib/services/transaction-manager.js]: Documented here fef9183

File: spec/integration/product-manager.spec.js (hunk 4 above)
Comment (lojzatran) [reply spec/integration/product-manager.spec.js]: I think we're covering all edge cases in our wiki. Or do you think there are cases that we didn't cover? Let's leave the test here but maybe we don't need more integration tests.

Comment (junajan) [reply spec/integration/product-manager.spec.js]: I think that in our examples we did not cover a lot of possible cases. But .. it does not make sense to test them in examples because we would make a lot of integration tests which are doing the same steps just with few small differences (like removing one variant or removing one variant and promoting another to masterVariant).

This can be tested with higher efficiency in productManager integration/unit tests.

Also the more functions we add to productManager (eg: anonymizing variants will be the next) the more combinations you have so you should test rather separated methods instead of the whole process for all possible use cases. It will be much faster (you won't test same methods over and over) and much easier for future changes (when you change one method you will not have to change it in all tests). Also, the tests will be small because they target just small pieces of the code instead of big tests which tests whole process.
